PSOperations/Capability.swift

```diff
@@ -8,15 +8,53 @@
 
 import Foundation
 
-public enum CapabilityErrorCode: Int {
-    public static var domain = "CapabilityErrors"
+public struct CapabilityError: Error, Equatable {
+    public let capabilityName: String
+    public let reason: Reason
     
-    case notDetermined
-    case notAvailable
-    case denied
+    public init<Capability: CapabilityType>(capability: Capability, reason: Reason) {
+        self.capabilityName = Capability.name
+        self.reason = reason
+    }
+    
+    public init?<Capability: CapabilityType>(capability: Capability, status: CapabilityStatus) {
+        guard let reason = status.errorReason else { return nil }
+        self.capabilityName = Capability.name
+        self.reason = reason
+    }
+}
+
+public extension CapabilityError {
+    public enum Reason: Equatable {
+        case notDetermined
+        case notAvailable
+        case denied
+        case failed(Error)
+        
+        public static func ==(lhs: Reason, rhs: Reason) -> Bool {
+            switch (lhs, rhs) {
+                case (.notDetermined, .notDetermined),
+                     (.notAvailable, .notAvailable),
+                     (.denied, .denied):
+                    return true
+            case (.failed(_ /*let lhsError*/), .failed(_ /*let rhsError*/)):
+                return true
+                // We could compare based on the errors bridged to NSError. But it would be rather unsafe.
```

Comment:
Why not make a separate case just for this?
```
        case (.failure(let l as NSError), .failure(let r as NSError)):
            return l == r

        case (.failure, .failure):
            return true
```
Alternatively you could make the Error generic
```
enum Reason<E: Error> {
    case notDetermined
    case notAvailable
    case denied
    case failed(E)
}

extension Reason: Equatable where E: Equatable {
    static func ==(lhs: Reason, rhs: Reason) -> Bool {
        switch (lhs, rhs) {
        case (.notDetermined, .notDetermined),
             (.notAvailable, .notAvailable),
             (.denied, .denied):
            return true

        case (.failed(let l), .failed(let r)):
            return l == r

        default:
            return false
        }
    }
}
```

Reply:
Because all Swift `Error`s are bridgeable to `NSError`.

Reply:
If they are all bridgeable, they can all be equated, can they not?

Reply:
Because `NSError` does not e.g. use a custom `Equatable` implementation. It only checks the type.
```swift
struct MyCustomError: Error, Equatable {
    let someInt: Int
    let someString: String
}

let error1 = MyCustomError(someInt: 4, someString: "ABC")
let error2 = MyCustomError(someInt: 3, someString: "DEF")

error1 == error2 // false
(error1 as NSError) == (error2 as NSError) // true
```

Reply:
Resolved by removing the `Equatable` conformance.